cmd/soci/commands/rebuild_db.go

```diff
@@ -0,0 +1,155 @@
+		// check if artifact db contains all artifacts in content store, if not add them
+		err = filepath.WalkDir(blobStorePath, func(path string, d fs.DirEntry, err error) error {
+			if !d.IsDir() {
+				info, err := d.Info()
+				if err != nil {
+					return err
+				}
+				f, err := os.Open(path)
+				if err != nil {
+					return err
+				}
+				defer f.Close()
+				var sociIndex soci.Index
+				err = soci.DecodeIndex(f, &sociIndex)
+				// skip: file is either a ztoc or empty config
+				if err != nil || info.Size() < 10 {
+					return nil
+				}
+				indexDigest := addHashPrefix(d.Name())
+				ae, err := artifactsDb.GetArtifactEntry(indexDigest)
+				if err != nil && !errors.Is(err, soci.ErrArtifactBucketNotFound) && !errors.Is(err, errdefs.ErrNotFound) {
+					return err
+				}
+				if ae == nil {
+					manifestDigest := sociIndex.Subject.Digest.String()
+					img, err := client.ImageService().List(ctx, fmt.Sprintf("target.digest==%s", manifestDigest))
+					if err != nil {
+						return err
+					}
+					spec, err := images.Platforms(ctx, client.ContentStore(), img[0].Target)
+					if err != nil {
+						return err
+					}
+
+					indexEntry := &soci.ArtifactEntry{
+						Size:           info.Size(),
+						Digest:         indexDigest,
+						OriginalDigest: manifestDigest,
+						ImageDigest:    manifestDigest,
+						Platform:       platforms.Format(spec[0]),
+						Type:           soci.ArtifactEntryTypeIndex,
+						Location:       manifestDigest,
+						MediaType:      sociIndex.MediaType,
+						CreatedAt:      time.Now(),
+					}
+					err = artifactsDb.WriteArtifactEntry(indexEntry)
+					if err != nil {
+						return err
+					}
+					for _, zt := range sociIndex.Blobs {
+						ztocEntry := &soci.ArtifactEntry{
+							Size:           zt.Size,
+							Digest:         zt.Digest.String(),
+							OriginalDigest: zt.Annotations[soci.IndexAnnotationImageLayerDigest],
+							Type:           soci.ArtifactEntryTypeLayer,
+							Location:       zt.Annotations[soci.IndexAnnotationImageLayerDigest],
+							MediaType:      soci.SociLayerMediaType,
+							CreatedAt:      time.Now(),
+						}
+						err := artifactsDb.WriteArtifactEntry(ztocEntry)
+						if err != nil {
+							return err
+						}
+					}
+				}
+
+			}
+			return nil
+		})
+		if err != nil {
+			return err
+		}
+		// check if artifact db contains artifacts that no longer exist in content store, if so delete them
+		err = artifactsDb.Walk(func(ae *soci.ArtifactEntry) error {
+			existsInContentStore, err := blobStore.Exists(context.Background(), v1.Descriptor{MediaType: ae.MediaType, Digest: digest.Digest(ae.Digest)})
+			if err != nil {
+				return err
+			}
+			if !existsInContentStore {
+				err := artifactsDb.RemoveArtifactEntryByDigest(ae.Digest)
+				if err != nil {
+					return err
+				}
+			}
+			return nil
+		})
+		return err
+	},
+}
+
+// addHashPrefix adds either a sha256 or sha512 prefix to a string
+func addHashPrefix(name string) string {
```

Comment:
I think this is what `godigest.NewDigestFromBytes` is for: https://pkg.go.dev/github.com/opencontainers/go-digest#NewDigestFromBytes

Reply:
We already have the hash encoding as its the filename. We just need to determine the hash algorithm (sha256 vs 512) and we can do this obtaining the length of the encoding.